packages/zwave-js/src/lib/commandclass/PowerlevelCC.ts

```diff
@@ -0,0 +1,409 @@
+		min: 1,
+		max: 255,
+		label: "Timeout",
+		description: "Time in seconds before node returns to Normal power.",
+	})
+	public readonly timeout?: number;
+
+	public toLogEntry(): MessageOrCCLogEntry {
+		return {
+			...super.toLogEntry(),
+			message: { "Power level": this.powerlevel, Timeout: this.timeout },
+		};
+	}
+}
+
+@CCCommand(PowerlevelCommand.Get)
+@expectedCCResponse(PowerlevelCCReport)
+export class PowerlevelCCGet extends PowerlevelCC {}
+
+interface PowerlevelCCTestNodeSetOptions extends CCCommandOptions {
+	testNodeId: number;
+	powerlevel: Powerlevel;
+	testFrameCount: number;
+}
+
+@CCCommand(PowerlevelCommand.TestNodeSet)
+export class PowerlevelCCTestNodeSet extends PowerlevelCC {
+	public constructor(
+		driver: Driver,
+		options:
+			| CommandClassDeserializationOptions
+			| PowerlevelCCTestNodeSetOptions,
+	) {
+		super(driver, options);
+		if (gotDeserializationOptions(options)) {
+			// TODO: Deserialize payload
+			throw new ZWaveError(
+				`${this.constructor.name}: deserialization not implemented`,
+				ZWaveErrorCodes.Deserialization_NotImplemented,
+			);
+		} else {
+			const testNode = driver.controller.nodes.get(options.testNodeId);
+			if (!testNode) {
+				throw new ZWaveError(
+					`${this.constructor.name}: Node Id ${options.testNodeId} doesn't exist.`,
+					ZWaveErrorCodes.Argument_Invalid,
+				);
+			}
+			if (testNode.isFrequentListening) {
+				throw new ZWaveError(
+					`${this.constructor.name}: Node Id ${options.testNodeId} is FLiRS, which is not allowed for the Test Node.`,
+					ZWaveErrorCodes.Argument_Invalid,
+				);
+			}
+			if (testNode.status !== NodeStatus.Awake) {
+				throw new ZWaveError(
+					`${this.constructor.name}: Node Id ${options.testNodeId} status is ${testNode.status}, but the Test Node must be Awake.`,
+					ZWaveErrorCodes.Argument_Invalid,
+				);
+			}
+
+			this.testNodeId = options.testNodeId;
+			this.powerlevel = options.powerlevel;
+			this.testFrameCount = options.testFrameCount;
+		}
+	}
+
+	public testNodeId: number;
+	public powerlevel: Powerlevel;
+	public testFrameCount: number;
+
+	public serialize(): Buffer {
+		this.payload = Buffer.from([
+			this.testNodeId,
+			this.powerlevel,
+			(this.testFrameCount >> 8) & 0xff,
+			this.testFrameCount & 0xff,
+		]);
+		return super.serialize();
+	}
+
+	public toLogEntry(): MessageOrCCLogEntry {
+		return {
+			...super.toLogEntry(),
+			message: {
+				testNodeId: this.testNodeId,
+				powerlevel: this.powerlevel,
+				testFrameCount: this.testFrameCount,
+			},
+		};
+	}
+}
+
+@CCCommand(PowerlevelCommand.TestNodeReport)
+export class PowerlevelCCTestNodeReport extends PowerlevelCC {
+	public constructor(
+		driver: Driver,
+		options: CommandClassDeserializationOptions,
+	) {
+		super(driver, options);
+
+		validatePayload(this.payload.length >= 4);
+		this.testNodeId = this.payload[0];
+		this.testStatus = this.payload[1];
+		this.testFrameAcknowlegedCount =
+			(this.payload[2] << 8) | this.payload[3];
```

Comment:
Use `readUInt16BE` instead

Reply:
> > All of the report values are non-stateful. Hopefully I handled those correctly.
> 
> I just read this after the review. In this case, remove all the @ccValue and @ccValueMetadata stuff to make sure the values aren't stored automatically.

Doesn't `@ccValue({ stateful: false })` cause the value to be emitted as a "value notification" and not stored in the DB?  I do see that having all of the separate properties will cause multiple events, each with just one part of the result.

What I really want to do is have `PowerlevelCCReport` produce one "value notification" with `timeout` as the value and `powerlevel` in the label.

`PowerlevelCCTestNodeReport` is a bit less obvious to implement.  It should produce one "value notification" with `testFrameAcknowlegedCount` as the value and `testStatus`, `testFrameCount`, `powerlevel`, and `testNodeId` imbedded as metadata.  The `testStatus` value would be indexed by the `testNodeId` (like Central Scene indexes values by `sceneNumber`). I believe that I can accomplish that with `ccSpecific` metadata.

> A few pointers in terms of the API:
> I'd rename `startTestNode` -> `startNodeTest` and `getTestNodeResult` to `getNodeTestStatus` (since it can be ongoing).
> I agree that the `TestNodeSet` shouldn't wait for the unsolicited report. This functionality would best be offered through a method on `ZWaveNode`, that returns a Promise which resolves when the report comes in.

Does it make sense to offer this method in addition to generating the "value notifications"?  I think I can see use cases that would prefer both options.  The `ZWaveNode` method would be simpler for a client that just wants to start the test and await the result.  The "value notifications" would be really easy to integrate into Home Assistant; the user would just need to kick off a service call and listen on the `zwave_js_event` for the matching "value notification".

Reply:
I can see your point, but I prefer having values only for CCs that a user often interacts with. I'd probably remove the SET/POLL_VALUE api completely for this CC. Such management commands are better suited for specialized methods on the node/controller/driver. Examples are the network heal, firmware update.

The API should make sense standalone and not try to work around how HA works internally.